petastorm/unischema.py

```diff
@@ -401,19 +401,41 @@ def insert_explicit_nulls(unischema, row_dict):
                 raise ValueError('Field {} is not found in the row_dict, but is not nullable.'.format(field_name))
 
 
+def _fullmatch(regex, string, flags=0):
+    """Emulate python-3.4 re.fullmatch()."""
+    if six.PY2:
+        m = re.match(regex, string, flags=flags)
+        if m and (m.span() == (0, len(string))):
+            return m
+    else:
+        return re.fullmatch(regex, string, flags)
+
+
 def match_unischema_fields(schema, field_regex):
     """Returns a list of :class:`~petastorm.unischema.UnischemaField` objects that match a regular expression.
 
     :param schema: An instance of a :class:`~petastorm.unischema.Unischema` object.
-    :param field_regex: A list of regular expression patterns.
+    :param field_regex: A list of regular expression patterns. A field is matched if the regular expression matches
+      the entire field name.
     :return: A list of :class:`~petastorm.unischema.UnischemaField` instances matching at least one of the regular
       expression patterns given by ``field_regex``.
     """
     if field_regex:
         unischema_fields = []
+        legacy_unischema_fields = []
         for pattern in field_regex:
             unischema_fields.extend(
-                [field for field_name, field in schema.fields.items() if re.match(pattern, field_name)])
+                [field for field_name, field in schema.fields.items() if _fullmatch(pattern, field_name)])
```

Comment:
Nit: why are we using `.extend()` here? Could we just say `legacy_unischema_fields =  [field for field_name, field in schema.fields.items() if _fullmatch(pattern, field_name)]`?

Reply:
We are going over a list of regular expression. For each one, we need extend the `unischema_fields` list. In your code samples, we would end up constantly overwriting it and use only the last one.